### PR TITLE
Example on Financial Time Series Imputation

### PR DESCRIPTION
**Pull request by kayuksel**
[dataset.csv.zip](https://github.com/kayuksel/torchgan/files/2987327/dataset.csv.zip)

Pull request discussed over #90

### CODE REVIEW

File: examples/fx_gan.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        super(base_conv1d, self).__init__()
+        self.conv = nn.Conv1d(in_channels, out_channels, 
+            bias = False, kernel_size = ks, padding=pd, **kwargs)
+        self.activation = nn.RReLU(inplace = True)
+        self.conv.weight.data.normal_(0, 0.01)
+    def forward(self, x):
+        return self.activation(self.conv(x))
+
+# Discriminator needs to take the generator output or the real sample and guess whether fake or not
+class Discriminator(nn.Module):
```
Comment (avik-pal) [inline examples/fx_gan.py]: Discriminator should inherit from the `models.Discriminator`.

File: examples/fx_gan.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        self.conv2 = base_conv1d(in_channels = 512, out_channels = 256, ks = 5, pd = 2)
+        self.conv3 = base_conv1d(in_channels = 256, out_channels = 128)
+        self.out = out_layer(in_channels = 128, out_channels = out_channels, gap_size = gap_size)
+    def forward(self, x):
+        if self.training and self.noise != 0.0:
+            x = x + Variable(x.data.new(x.size()).normal_(0.0, self.noise))
+        return self.out(self.conv3(self.conv2(self.conv1(x))))
+
+# Generator needs to take the latent variables and then generate a fake sequence of given channels
+class Generator(nn.Module):
```
Comment (avik-pal) [inline examples/fx_gan.py]: Same here make it inherit `models.Generator`

File: examples/fx_gan.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        self.conv1 = base_conv1d(in_channels = 128, out_channels = 256)
+        self.conv2 = base_conv1d(in_channels = 256, out_channels = 512, ks = 5, pd = 2)
+        self.conv3 = base_conv1d(in_channels = 512, out_channels = feat_length-input_size, ks = 7, pd = 3)
+        self.conv3.activation = nn.Sigmoid() # Because the input is in [0, 1]
+    def forward(self, x, direction = True):
+        gen = Variable(torch.randn(x.shape[0], 128, self.channels)).cuda()
+        gen = self.conv3(self.conv2(self.conv1(gen))).view(-1, self.channels, feat_length-input_size)
+        return torch.cat([gen, x], dim=2) if direction else torch.cat([x, gen], dim=2)
+
+Dataset = pd.read_csv('Dataset.csv', index_col='Gmt time', parse_dates=True, infer_datetime_format=True)
```
Comment (avik-pal) [inline examples/fx_gan.py]: If we are going to have this in the examples the dataset should be downloadable from somewhere and we should have proper license to do so. If this one doesn't conform to these, we have to use some other dataset.

File: examples/fx_gan.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+train_loader = torch.utils.data.DataLoader(train_dataset, 
+    batch_size = 64, shuffle = True, pin_memory = True)
+
+G = Generator(channels = feat_channel, gap_size = feat_length).cuda()
+D = Discriminator(channels = feat_channel, gap_size = feat_length).cuda()
+
+G_opt = torch.optim.Adam(params = G.parameters(), lr = 1e-3)
+D_opt = torch.optim.Adam(params = D.parameters(), lr = 1e-3)
+
+BCE = torch.nn.BCELoss()
```
Comment (avik-pal) [inline examples/fx_gan.py]: Use the `losses.functional` API to create the Generator and Discriminator Loss.

File: examples/fx_gan.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+G.train(True)
+D.train(True)
+
+for epoch in range(epochs):
```
Comment (avik-pal) [inline examples/fx_gan.py]: This train loop is pretty trivial and can be handled directly using the `Trainer`. You just need to define the models and losses as I have described then feed them to the Trainer. You don't need to implement any plotting explicitly those are already taken care of.
